ca/crl.go

```diff
@@ -0,0 +1,164 @@
+package ca
+
+import (
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"time"
+
+	capb "github.com/letsencrypt/boulder/ca/proto"
+	"github.com/letsencrypt/boulder/core"
+	"github.com/letsencrypt/boulder/issuance"
+	blog "github.com/letsencrypt/boulder/log"
+)
+
+type crlImpl struct {
+	capb.UnimplementedCRLGeneratorServer
+	issuers  map[issuance.IssuerNameID]*issuance.Issuer
+	lifetime time.Duration
+	log      blog.Logger
+}
+
+func NewCRLImpl(issuers []*issuance.Issuer, lifetime time.Duration, logger blog.Logger) (*crlImpl, error) {
+	issuersByNameID := make(map[issuance.IssuerNameID]*issuance.Issuer, len(issuers))
+	for _, issuer := range issuers {
+		issuersByNameID[issuer.Cert.NameID()] = issuer
+	}
+
+	if lifetime == 0 {
+		logger.Warningf("got zero for crl lifetime; setting to default 9 days")
+		lifetime = 9 * 24 * time.Hour
+	} else if lifetime >= 10*24*time.Hour {
+		return nil, fmt.Errorf("crl lifetime cannot be more than 10 days, got: %s", lifetime)
+	} else if lifetime <= 0*time.Hour {
+		return nil, fmt.Errorf("crl lifetime must be positive, got: %s", lifetime)
+	}
+
+	return &crlImpl{
+		issuers:  issuersByNameID,
+		lifetime: lifetime,
+		log:      logger,
+	}, nil
+}
+
+func (ci *crlImpl) GenerateCRL(stream capb.CRLGenerator_GenerateCRLServer) error {
+	rcs := make([]pkix.RevokedCertificate, 0)
+	var number *big.Int
+	var thisUpdate time.Time
+	var issuer *issuance.Issuer
```

Comment:
It was a little surprising to see these declarations hanging up so far from their use. What if we just store a `*capb.GenerateCRLRequest_Metadata` object outside the loop, and do the processing (getting issuer, making bigint, etc) once the loop exits? That reduces our ability to early-exit for unknown issuers, but it shortens the loop and lets us reduce predeclaration. This would also let us use a nil check to figure out "did we already get metadata?"

We could also do much of the `capb.GenerateCRLRequest_Entry` branch in its own function that produces a pkix.RevokedCertificate.